### ISSUE TITLE
Could the releases on github be simplified / reduced?

### ISSUE THREAD

**Issue by rubyFeedback** [Alex313031/thorium#716]
Right now we have this here:

    thorium-browser_124.0.6367.218_AVX.deb
    thorium-browser_124.0.6367.218_AVX.rpm
    thorium-browser_124.0.6367.218_AVX.zip
    thorium-browser_124.0.6367.218_AVX2.deb
    thorium-browser_124.0.6367.218_AVX2.rpm
    thorium-browser_124.0.6367.218_AVX2.zip
    thorium-browser_124.0.6367.218_SSE3.deb
    thorium-browser_124.0.6367.218_SSE3.rpm
    thorium-browser_124.0.6367.218_SSE3.zip
    thorium-browser_124.0.6367.218_SSE4.deb
    thorium-browser_124.0.6367.218_SSE4.rpm
    thorium-browser_124.0.6367.218_SSE4.zip
    Thorium_Browser_124.0.6367.218_AVX.AppImage
    Thorium_Browser_124.0.6367.218_AVX2.AppImage
    Thorium_Browser_124.0.6367.218_SSE3.AppImage
    Thorium_Browser_124.0.6367.218_SSE4.AppImage
    Source code (zip)
    Source code (tar.gz)

18 entries. I am not entirely certain all of these are necessary, but, ignoring this, I find the differences confusing:

AVX, AVX2, SSE3, SSE4.

I have no idea what to download now. I remember this used to be simpler.

I don't know what any of these abbreviations mean either. I may download the wrong thing.

Could either:

a) everything be simplified to just download ONE variant?
b) if a) is not possible, then a good explanation be added, be it on the README or the wiki or some other site,
and then link towards that explanation?

Right now I have no idea what to download anymore.

A quick google showed SSE3 is SSE3, Streaming SIMD Extensions 3, SSE4 is SSE4 (Streaming SIMD Extensions 4),
AVX2 is something about vectors - but none of that tells me anything what I should download.

**Comment by gz83** [comment]
In general, it is recommended to choose the version with the highest instruction set supported by the processor. In addition, due to user requirements, we have released multiple different versions and cannot merge them into one.

The description of the different versions may be lacking, but I don't think it should cause too much confusion.

**Comment by TetraTheta** [comment]
SSE3/4, AVX/2 are CPU instruction set. If your CPU supports _newer_ instruction and you use build that compiled for it, you'll get performance gain from it. In general, AVX is newer instruction set than SSE.

You can inspect which instruction set your CPU supports by using software like CPU-Z.

**Comment by Kinuseka** [comment]
AVX2 is usually preferred for newer processors (2015+)

If you're running on older processors SSE will run just fine.

**Comment by anonyco** [comment]
Looking for this! And I want to add my 5 cents as a low-level C software dev who frequently deals with multiversioning code to different ISA support levels for performance reasons and inspects compiled assembly.

Currently, Thorium has pitiful architecture support features that neither reflect the market share demographics nor the optimum features for performance. I'd imagine a lot more people will benefit from a much faster Thorium once its versions are simplified into 4 separate groupings that will give >=80% of optimal hardware utilization for >=80% of users (ballpark numbers based on my experience):

1.  **2008-2015 CPUs with SSE4.1 build for Intel Broadwell/earlier and AMD Bobcat 2 and all AMD Bulldozer** (enables `SSE4.1` and earlier,  `CX16` and `POPCNT`, which all SSE4.1-supporting CPUs also support): arguably a larger demographic than old PCs is emulating x86_64 software on non-x86_64 architectures in virtual machines. Practically every desktop/laptop computer architecture system in the past 20 years has has *some* kind of SIMD instruction set, and the overwhelming majority lend themselves as a not-terrible/sometimes-decent emulation target for SSE4.1 x86_64 SIMD instructions and earlier. SSE4.2 string instructions are horribly inefficient to emulate on non-x86_64, which is why SSE4.1 is the sweet spot of sorts supporting all kinds of misc users/usecases without forcing users to downgrade all the way to baseline SSE/SSE2 (which is missing a significant chunk of very important instructions, e.g. lack of pshufb greatly limits what algorithms can be vectorized, while lack of roundsd causes ceil/floor floating point calls in C to become expensive libc function calls and greatly slows down a lot of floating point code.)
    - Also includes **2016-2023 Intel Celeron/Pentium/Atom lowest-end CPUs** (much more common in international markets)
3. **2016-2026 CPUs with AVX2+AES+CLFLUSHOPT build for E-Core/P-Core Intel CPUs without AVX10.2 and AMD Zen1/Zen2/Zen3 and various Skylake/newer without AVX512** (enables `SSE4.2` and earlier, `'POPCNT`, `CX16`, `PCLMUL`, `FSGSBASE`, `RDRND`, `F16C`, `AVX2` , `BMI2`, `FMA`, `MOVBE`, `RDSEED`, `ADCX`, `PREFETCHW`, `AES`, and `CLFLUSHOPT`). Each of those enabled instructions does give appreciable speedups to certain algorithms/code, specifically you're guaranteed a bad day web browser without AES (which is super slow to do in software) and BMI2 allows compilers to emit (sometimes significantly) shorter instruction sequences for multiplication, bitshifts, and other bitops that speed up most-all code everywhere. (Note: AES hardware instructions are a big enough deal its certainly run-time detected, however the benefits still stand.) Note: no mainstream Intel CPU has AVX10.1 yet; if the CPU has E-cores and its info page doesn't list AVX10 use this build. For future E-core CPUs with AVX10.1, use the next No. 4 build option. 
4. **2016-Present CPUs with AVX10.1/weak-AVX512 build for early AVX512 CPUs Skylake-AV512 through Rocketlake and AMD Zen4** (enables `-mavx10.1` plus all features from No 3 last build). Basically, the reason AVX512 boosts performance is long-overdue instructions that really should have been introduced a decade ago with AVX/2, e.g. encoding explicit rounding modes means libc math routines dont have to painfully save/reload the mxcsr *every* single *function* call (*yes,* its *really* this *bad*). The 512-bit aspect of AVX512 actually nets negligible performance gains over AVX2 as 256 bit vectors are *wide enough* to efficiently handle/swizzle the overwhelming majority of algorithms. Moreover, a lot of early AVX512 CPUs have only 1 AVX512 ports but 2 or 3 AVX256/128 ports and/or double-pump AVX512 like Znver4, meaning fully utilizing the CPU requires custom hand-tuning that's not possible for a compiler to interleave 512-bit and 256-bit operands, yet only net a 33% (rarely 50% at most) performance boost over all-256-bit vectors or all 512-bit vectors. Moreover, even if a majority of critical code were hand optimized for the 33-50% boost (which its not), the remaining non-vectorized code dominates execution time enough in "typical" software like this web browser to mean you'd be unlikely to even measure the actual speedup if any. SO, the solution is lump all these seemingly unrelated CPU targets into one AVX10.1 build, which exactly equates building for AVX512 EXCEPT disabling/preventing 512-bit zmm vector registers. AVX10.1 adds no new instructions and is compatible with all CPUs that support full proper AVX512, running efficiently on those AVX512 CPUs. (Yes, there's rumors of AVX10.1 supporting 512-bit; it likely won't in the real CPUs.) Again with the 33-50% boost to hand-tuned SIMD code for 512-bit vector registers: AVX10.1 is a very justifiable grab-all target that won't harm performance for any AVX512 CPU (versus the performance of compiling with full AVX512 512-bit vectors) and, in some cases like Skylake-AVX512, is likely to boost performance as it avoids down-clocking the cpu.
5. **2023-Present maximum sever-grade AVX512 build for Zen5/newer and Xenon SapphireRapids/newer:** (which uses `-march=znver5 -mtune=znver5 -mno-waitx -mno-clzero -mno-avx512vp2intersect -mno-prefetchi` ONLY, no previous build flags.) First, notice the `-march=znver5` is a misnomer and it doesn't actually compile *exclusively* for AMD Zen5&mdash;the `-mno-` ensure it also works on Sapphirerapids too. Basically, Zen5 blows all other CPUs out of the water. The 7% IPC gains over Zen4 some benchmarks report and the modest/humble 20% given by AMD do no justice to the ridiculous power and performance and Zen5. Some algorithms can see as much as a 6-7x speedup over Zen4 raw-number-crunching-throughput-wise or a 2-3x speedup over the best pre-Sapphire-Rapids Intel CPU (and 1.3x-1.5x over Sapphire Rapids and newer throughput-wise.) The problem is realizing this performance. This is no fault of AMD and is no issue in Zen5's design--I'd be surprised if much gets better with Zen6--but an issue how software is written optimized for typical pre-Zen5 workloads and spends extra instructions to get around historically-expensive intructions like division, multiplication, vpmullq, bsf/bsr, and bunch of random special-purpose instructions. Zen5 dedicates the die area to chow down these expensive operations like a champ and handles 3-5 multiplications per cycle, 4 bsr/bsf, 3 vpmullq, and similar stats of misc. This makes software written to minimize use of those instructions bottleneck on dependency chains and starve Zen5 of its potential IPC. (However, if software optimized for Zen5+ were ran on Zen4/earlier or Alderlake/earlier, it'd run a lot slower than the way it is written now as it'd bottleneck on stack/memory access, frontend issue, and port saturation from more expensive instructions; that's why separate writing/optimization is needed for Zen5+.) Again, this is no fault of AMD: they're pushing the limits of silicone and its ridiculous the Zen5 IPC they managed to cram onto the die without inflating core sizes too much; moreover, this is a new trend Intel is following with sapphire rapids. SO, we basically need two versions of each software, one written for slow, limited pre-Zen5/pre-SapphireRapids, and one written for post-Zen5/post-SapphireRapids. Trying to run pre-Zen5 on Zen5 hardware gives you exactly what the benchmarks to date show: a paltry 7-20% IPC increase. Likewise, trying to run post-Zen5 software on pre-Zen5 hardware gives the opposite performance penalty--a non-negligible 7-20% performance hit. At the same time, offering a specialized post-Zen5 version for Zen5+ hardware can give as much as a 6-7x performance boost over Zen4 in parts of the software tuned for it. 

**Notice: if you are confused about the last three points, how they differ, and how to tell which one your computer falls into, that's normal!** Intel (and to a lesser extent AMD) have radically changed CPU architectures enough over the past few years, which, coupled with Intels marketing blunders and AMD's double-pumping of AVX512 on Znver4, create a very complex and confusing landscape for SIMD performance optimization. The most egregious brain-dead Intel clusterfuck was shipping Alderlake with AVX512 on the P-cores (which anybody with half a braincell could have foreseen doesn't work at the OS/software level; AVX512 is all-or-none) THEN retroactively shipping firmware updates to disable AVX512 AND NOW backstepping on that decision of AVX512 on E/P cores with AVX10.1, which doesn't add any new assembly instructions but indicates feature bit support for non-512-bit-AVX512 on E-cores  (which anybody with half a braincell would have seen as the obvious solution BEFORE Alderlake was shipped and avoided this mess.)  Thorium really needs to make a list of all cpu architectures and which they fall into based on the [GCC manual](https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html#index-mvect8-ret-in-mem). However, as confusing as it is, I can tell you from my experience and the low level SIMD I've done that all you need is those 4 targets and those 4 targets only to get >=80% of optimal hardware utilization from >=80% of users.

(**Note 1:** If there is any user out there that *really* wants to run Thorium x86_64 on a non-x86_64 architecture via a VM AND their VM for whatever reason only supports baseline SSE and SSE2, not SSE4.1, then they can always use a different project other than Thorium (or, better yet, get native builds of plain Chromium for their architecture, which are guaranteed to run significantly better than trying to bootstrap Thorium via a VM as Thorium's optimizations are a drop in the bucket compared to the overhead of cross-architecture virtualization.) And, I seriously doubt such a user even exists.)

(**Note 2:** Plain AVX without AVX2 is mostly useless for general vector routines due to the lack of integer and swizzle operators. The only software that benefits from plain AVX over SSE4.1 is scientific software doing huge floating point calculations. String operations like memcpy can also benefit from AVX by using the twice-size vector regisers to move twice as much data at once, HOWEVER these benefits are only seen manipulating gigabytes of data AND the system library should automatically choose AVX-optimized string operations (as GCC and Clang turn huge/unknown string operations into libc calls, which further negates any benefits trying to enable it in Thorium.) Additionally, AVX can even run slower than SSE4.1 due to the overhead of longer instructions encoding vex prefixes and putting vzeroupper at the end of functions. I'd attribute any performance boost compiling for AVX over compiling for SSE4.1 to issues in the toolchain. (Track these issues down, fix them, and you'll see enough of a speedup in SSE4.1 its a negligible difference from AVX.))

**TL;DR:** Thorium needs to stop messing around with its crapshoot overcomplicated selection of support levels that woefully underutilizes the majority of users' hardware and simplify everything to the 4 options/explanations I bolded to achieve >=80% hardware utilization for >=80% of users.

**Comment by leao-nardo** [comment]
i understand that too many options seems "confusing/complicated" . and I'm ok with people that dont want options not having them. but when those people want that no one have options...well this bothers me. because i do want options and many ones do too.

and, actually, the whole point of this project is to make a fork of chromium optimized(or at least compatible) to specific architectures/instructions set .
 the man started building for speed in a bulldozer, if i remember, and then all the others that have now. 

i, inclusive, came here today for a sse3 build for an old atom laptop that still work. many laptop where made with those atom processors(in 2010 and around) and they dont have avx and not even sse4(look https://valid.x86.fr/wgptcn
https://cpu.userbenchmark.com/SpeedTest/13876/IntelR-AtomTM-CPU-N470-----183GHz#Specifications)_. this repo is exactly probably most used by people looking for those specifics architectures/instructions set builds. because if you compare it to any major fork, always, for years, this one lack behind a lot in keeping update with the current chromium(by months), what it does very well is to serve a quite maintained compiled builds for those specific old sets .

and i understand that your point/concern is, actually, more for not having an explanation on what they are and what to choose, and for that, past post here from @anonyco explains very well and the next ones having that doubts can find and read here.
